### PR TITLE
[DirectX] introducing lowering for `bufferUpdateCounter`

### LINKED ISSUE #92147: [DirectX] Introduce llvm.dx.updateCounter intrinsic and lower it to the dxil op
We need the intrinsic lowering for the [bufferUpdateCounter](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#bufferupdatecounter) DXIL op, which is emitted to implement [RWStructuredBuffer](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/sm5-object-rwstructuredbuffer)'s increment and decrement counter methods as well as `Append`/`Consume` methods on [AppendStructuredBuffer](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/sm5-object-appendstructuredbuffer) and [ConsumeStructuredBuffer](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/sm5-object-consumestructuredbuffer).

This is needed for #92148

https://godbolt.org/z/qKGdYj5zj
```
struct MyStruct {
    float4 a;
};

AppendStructuredBuffer<MyStruct> AB;
ConsumeStructuredBuffer<MyStruct> CB;

[numthreads(4,1,1)]
void main() {
  MyStruct val = CB.Consume();
  AB.Append(val);
}
```

### PR DESCRIPTION
**Pull request by joaosaffran**
- Adding custom lowering for `bufferUpdateCounter`
- introduces llvm intrinsic `int_dx_updateCounter`
- adds tests

Closes #92147

### CODE REVIEW

File: llvm/test/CodeGen/DirectX/updateCounter.ll [hunk 1]
```diff
@@ -0,0 +1,31 @@
+; RUN: opt -S -dxil-op-lower %s | FileCheck %s
```
Comment (bogner) [inline llvm/test/CodeGen/DirectX/updateCounter.ll:1]: I don't see any `CHECK:` lines in this test...

File: llvm/lib/Target/DirectX/DXIL.td [hunk 2]
```diff
@@ -730,6 +730,15 @@ def BufferStore : DXILOp<69, bufferStore> {
   let stages = [Stages<DXIL1_0, [all_stages]>];
 }
 
+def UpdateCounter : DXILOp<70, bufferUpdateCounter> {
+  let Doc = "increments/decrements a buffer counter";
+  let arguments = [
+    HandleTy, Int8Ty
+  ];
```
Comment (bogner) [inline llvm/lib/Target/DirectX/DXIL.td:737]: No need to wrap here, this list is short.
```suggestion
  let arguments = [HandleTy, Int8Ty];
```

**Review by bogner — APPROVED** [review]

File: llvm/include/llvm/IR/IntrinsicsDirectX.td [hunk 3]
```diff
@@ -35,6 +35,9 @@ def int_dx_typedBufferLoad_checkbit
 def int_dx_typedBufferStore
     : DefaultAttrsIntrinsic<[], [llvm_any_ty, llvm_i32_ty, llvm_anyvector_ty]>;
 
+def int_dx_updateCounter
+    : DefaultAttrsIntrinsic<[], [llvm_any_ty, llvm_anyint_ty]>;
```
Comment (spall) [inline llvm/include/llvm/IR/IntrinsicsDirectX.td:39]: is there an llvm_i8_ty that could be used here for the 2nd arg? From DXIL.td it seems this is supposed to only accept i8? I see a use of llvm_i8_ty here:   https://github.com/llvm-mirror/llvm/blob/2c4ca6832fa6b306ee6a7010bfb80a3f2596f824/include/llvm/IR/IntrinsicsX86.td#L16

File: llvm/lib/Target/DirectX/DXILOpLowering.cpp [hunk 4]
```diff
@@ -463,6 +463,28 @@ class OpLowerer {
     });
   }
 
+  [[nodiscard]] bool lowerUpdateCounter(Function &F) {
```
Comment (inbelic) [inline llvm/lib/Target/DirectX/DXILOpLowering.cpp:466]: Sorry for the late review. Is this going to be a common pattern to create a handle and then use it as an arg? Maybe it is possible to make a general helper.

Comment (bogner) [reply llvm/lib/Target/DirectX/DXILOpLowering.cpp:466]: Most of the intrinsics that use a handle are related to loads, stores, and sampling - bufferUpdateCounter is a bit special in its simplicity. As you might see from the handling of typed buffer loads and stores, the translation isn't as straightforward there. @joaosaffran asked me about generalizing this now, but I think it's probably best to just do this one directly for now and generalize when we have more related logic in to make sure we're not generalizing in a way we'll need to undo later.

Comment (inbelic) [reply llvm/lib/Target/DirectX/DXILOpLowering.cpp:466]: Okay, sounds good 👍

**Comment by hekota** [comment]
@joaosaffran - `updateBufferCounter` needs to return `uint` with the pre-incremented or pre-decremented counter value: 

See https://godbolt.org/z/rGhWbcrY5
https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/sm5-object-rwstructuredbuffer-incrementcounter

**Comment by joaosaffran** [comment]
Will address this issue in a future PR @hekota, thanks for pointing it out.

**Pull request merged**